### PR TITLE
Grid interactivity: Allow blocks to be positioned in manual mode using drag and drop

### PR DESCRIPTION
**Pull request by noisysocks**
This PR adds the ability to explicitly position a grid item by dragging and dropping it to a cell.

In order to allow more precise manipulation of the grid, when switching to manual mode all grid items are fixed in place. They can then be moved around with drag and drop, with the block movers (in any direction) or by changing the values of column and row in the Dimensions sidebar.

Positioning a grid item is done by setting the style.layout.columnStart and style.layout.rowStart attributes. These were added in https://github.com/WordPress/gutenberg/pull/59483.

The drag and drop mechanism works by having a DropZone in each of the GridVisualizer's cells. There's some trickery here to work around the fact thatGridVisualizer needs pointer-events: none so that you can interact with the blocks that are underneath the visualiser. Pointer events are disabled on the cells until the user begins to drag a block, at which point we enable pointer events so that the drop zones work as expected.

Steps for testing:

1. In Gutenberg > Experiments, enable "Grid interactivity"
2. Add a Grid block to a post or template and add a few items inside it.
3. Set the Grid to Manual mode; adjust Columns and Rows as desired. 
4. Try dragging and dropping items from one cell to another.

https://github.com/WordPress/gutenberg/assets/8096000/99ed66aa-cd58-401d-91f1-b2c45cb1edce

Note that if a block is too big for a particular cell and there are not enough empty adjacent cells to fit it, the drop zone won't appear.

Also note that in this configuration it's possible to expand a grid item in a way that causes it to overlap with adjacent blocks (because all blocks in this mode have fixed column and row starts). This doesn't feel ideal.

### CODE REVIEW

**Comment by tellthemachines** [comment]
This is a promising start! 

Can we show the column/row start in the child layout panel for each block? 

After dragging and dropping a few blocks on an existing Grid, selecting the Grid block triggers an `Uncaught TypeError: Cannot destructure property 'columnStart' of '(intermediate value)' as it is undefined.`

Also should we disable dragging into an empty grid area altogether for auto grids? It's not working too well (I tried to drop a single span block into an empty place and it turned into a 3-col block 😅)

**Comment by tellthemachines** [comment]
Oh, just noticed something else: we might want to add an explicit default `grid-template-rows` for manual grids, otherwise when we set it to manual it collapses in on itself because the generated auto rows have 0 height:

Screenshot 2024-04-30 at 9 40 29 AM

Screenshot 2024-04-30 at 9 41 55 AM

**Comment by tellthemachines** [comment]
I've rebased and fixed a few things on this PR so it's in roughly testable form. From playing around with it, I think the biggest thing it's missing is the ability to drop into any place on a manual grid when the grid isn't selected. Currently, dropping into the grid from outside it will only show the in-between placeholder, so the item ends up falling into the first available space on the grid:


https://github.com/WordPress/gutenberg/assets/8096000/f8ed82be-2652-4757-954e-bc40582d28dc


The other thing that I'm not sure about is if we hijack the existing manual mode for this, how to still make available the existing manual mode. #60941 might help somewhat, but we also need to provide back compat and make sure the change isn't confusing for folks using the current grid controls.

Additional feedback and ideas are very welcome here!

**Comment by noisysocks** [comment]
> I've rebased and fixed a few things on this PR so it's in roughly testable form. From playing around with it, I think the biggest thing it's missing is the ability to drop into any place on a manual grid when the grid isn't selected. 

Good point. I think it would make sense if `GridVisualizer` appears when you're dragging over the grid block.

> Currently, dropping into the grid from outside it will only show the in-between placeholder, so the item ends up falling into the first available space on the grid:

Do the in-between placeholders make sense when the grid is in manual mode? I suppose they would work okay if we implement the functionality where `columnStart` and `rowStart` are automatically shuffled around when the block order is changed e.g. when you re-order a block in the list view.

> The other thing that I'm not sure about is if we hijack the existing manual mode for this, how to still make available the existing manual mode. https://github.com/WordPress/gutenberg/pull/60941 might help somewhat, but we also need to provide back compat and make sure the change isn't confusing for folks using the current grid controls.

Yep good flag. I think it makes sense to do that in tandem with this.

**Comment by tellthemachines** [comment]
OK I've updated this PR so that when grid items are moved around inside the grid, the markup is correspondingly reordered. This works whether the items are dragged and dropped, moved with the block movers or their row/column start values are updated in the sidebar. 

The code is pretty messy, but it seems to be working as expected.

https://github.com/WordPress/gutenberg/assets/8096000/60b37485-90cc-46c4-9c9a-5c2dae6e2a6c

Points of friction I noticed:
* block toolbar jumping around when moving blocks with the mover means you can't really move more than one position at a time without stopping and relocating the buttons
* particularly when moving with the mover or via the sidebar controls, blocks going under/over other blocks with no warning can be confusing

Still to be addressed, probably in follow-up PRs (this changeset is already pretty big)

* Show inserter in empty cells so we can add blocks to any position in the grid
* Allow dragging blocks from outside into the grid
* Resizing positioned blocks should re-position them according to the direction they were resized in

Any testing and feedback much appreciated!

File: packages/block-editor/src/components/block-mover/index.js
Comment (tellthemachines) [inline packages/block-editor/src/components/block-mover/index.js]: Can we disable block movers here without referring to layout at all?

File: packages/block-editor/src/components/child-layout-control/index.js
Comment (tellthemachines) [inline packages/block-editor/src/components/child-layout-control/index.js]: Welp, had to add a bunch of hooks here to be able to move the blocks in the markup order 😅 
Not sure if there's a better way of doing this. A possible alternative would be not exposing these controls in the sidebar at all (which would mean keyboard manipulation would depend exclusively on the block movers, maybe not ideal)

**Comment by jasmussen** [comment]
This is coming together. The fact that cells now highlight when you drop on them, and the way the markup reorients itself, feels much more predictable. Here's a GIF, showing both how this is better, and a few gotchas still:

grid after

* The sibling inserter is a point of confusion. It still appears, and seems to suggest places where you can drop the block you're dragging. Which is valid when you drag the item _outside of the grid_ where this is accurate. But this line also appears when you're dragging inside the grid, where it's actually inaccurate; you can't drop where the line suggests you can, and in some cases the line appears in what looks like the wrong place. If this line can be simply hidden whever it is inaccurate, that seems best.
* After resizing an item to have a colspan of 2, drag and drop had a slightly off behavior; it correctly highlights two cells when I can drop there. But if there isn't space, dropping there simply isn't possible.

I tried turning on to manual mode, which I have a feeling is eventually going to be a better default experience. Main reason: allowing overlapping containers is one of the most compelling aspects of a grid layout, and doesn't require you to fiddle with negative margins. It's also more predictable as far as resize handles and drag and drop; things stay in place. GIF showing manual mode:

grid after, manual mode

* The resizing to create overlap indeed works. But if I resize an item to span two columns, then move it to a free set of cells, I can't move it back. I should be allowed to create overlap, but it seems disallowed at the moment.
* The mover controls are hidden. But only after turning on manual mode, then deselecting a block and selecting a new one. I.e. if I have a cell block selected while turning on manual mode, I still have the movers. 
* There was some confusing undo behavior: I set an item to colspan 2, dragged it to another place, pressed "undo", expecting the colspan to remain 2 and only undoing the drag action. This appeared to also be the case in practice, however _visibly_ the item now looked as if it was colspan 1.

File: packages/block-editor/src/style.scss
Comment (andrewserong) [inline packages/block-editor/src/style.scss]: Just a quick drive-by comment: now that the `style.scss` has been moved back to `grid-visualizer`, looks like this line needs to be reverted, too?

Comment (tellthemachines) [reply packages/block-editor/src/style.scss]: oh well spotted, thanks! updated.

**Comment by tellthemachines** [comment]
Thanks for the comprehensive feedback @jasmussen!

> The sibling inserter is a point of confusion. It still appears, and seems to suggest places where you can drop the block you're dragging.

Agreed, and I've updated this branch to hide the sibling inserter when the grid is in manual mode, and show it only in auto. 

I've also hidden the cell highlight in auto mode, so blocks can only be positioned in manual mode. This allows for a clearer distinction between the two modes, and it also avoids many problems that arise when we have positioned elements in auto-reflowing grids 😅 

> if I resize an item to span two columns, then move it to a free set of cells, I can't move it back. I should be allowed to create overlap, but it seems disallowed at the moment.

This is also fixed now!

> The mover controls are hidden. But only after turning on manual mode, then deselecting a block and selecting a new one. I.e. if I have a cell block selected while turning on manual mode, I still have the movers.

I couldn't reproduce this so far, but will look into it further. 

Also have noted some weirdness with Undo, looking into that too.

I think once those bugs are sorted out we'll be in a reasonable place to merge this experiment. There are a few things still to do, but it might be easier on the reviewers if we leave them for follow-up PRs:

* Allow dragging blocks from outside the grid into a specific position.
* Add new blocks to any position with inserter in empty cells
* Resizing positioned blocks should re-position them when resizing top or left, so the block doesn't jump to a different position after resize.
* Handle grid items being re-arranged in list view (currently dragging and dropping in list view inside a manual grid causes markup to shift but blocks to visually remain in place)

**Comment by tellthemachines** [comment]
Ok I've fixed the undo issue, and still can't repro the movers issue. I think this is now ready for a code review.

**Comment by tellthemachines** [comment]
> Resizing positioned blocks should re-position them when resizing top or left, so the block doesn't jump to a different position after resize.

This is now also fixed!

File: packages/block-editor/src/components/block-mover/index.js
Comment (noisysocks) [reply packages/block-editor/src/components/block-mover/index.js]: Yeah, I can't really think of an alternative approach. 

Maybe if we had a more flexible way of disabling moving than the existing locking APIs. Then we could have the `layout` block-supports call an API to lock moving. But I don't think it's worth adding that just for this feature.

Should we use `useLayout` though instead of reading the attributes directly? I can't remember what benefit `useLayout` provides, if any.

File: packages/block-editor/src/components/grid-visualizer/grid-item-movers.js
Comment (noisysocks) [inline packages/block-editor/src/components/grid-visualizer/grid-item-movers.js:37]: We have a lot of `parseInt` in this PR. Is it necessary? Can't we guarantee that the attributes are numbers?

File: packages/block-editor/src/components/block-mover/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/block-mover/index.js]: `useLayout` doesn't return the correct layout type here; it only ever returns "default" in my testing. I don't think this component has access to the context from `BlockListItems`, which is what `useLayout` returns.
If it did work, the benefit would be accessing the parent layout type through context instead of a selector.

File: packages/block-editor/src/components/grid-visualizer/grid-visualizer.js
Comment (noisysocks) [inline packages/block-editor/src/components/grid-visualizer/grid-visualizer.js]: Oops this should be `__cell`.

**Comment by tellthemachines** [comment]
Update: I've started an experiment in #62777 to track manual/auto mode with a new attribute, so we can leverage both minimumColumnWidth and columnCount to make responsive grids in either mode.

File: packages/block-editor/src/components/grid-visualizer/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/grid-visualizer/index.js:4]: As a follow-up PR I definitely think we _should_ rename the folder from `grid-visualizer` to something like `grid` since components like this one have nothing to do with visualisation.

File: packages/block-editor/src/components/block-mover/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/block-mover/index.js]:
```suggestion
```

**Comment by noisysocks** [comment]
This is looking good. I refactored `GridVisualizer` and `useGetBlocksBeforeCurrentCell` a little to make things (I think) more readable. 

I'd like to take one more look at `useGridSync` then I think we can merge.

We should probably also do a last round of testing.

File: packages/block-editor/src/components/grid-visualizer/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/grid-visualizer/index.js:4]: Yes! Let's do that as a follow-up in its own PR, to avoid complicating reviews of actual code changes 😅

File: packages/block-editor/src/components/grid-visualizer/grid-item-movers.js
Comment (tellthemachines) [reply packages/block-editor/src/components/grid-visualizer/grid-item-movers.js:37]: When the values are coming in straight from the controls they're always strings, so this has to happen at some point.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/inner-blocks/index.js:274]: Bit difficult to understand what's going on here. Might be good to split the logic preceding the `?` into a variable e.g. `isDropZoneEnabled`.

File: packages/block-editor/src/components/grid-visualizer/grid-item-movers.js
Comment (noisysocks) [reply packages/block-editor/src/components/grid-visualizer/grid-item-movers.js:37]: Might be good to do an audit of this after this PR merges. Ideally we only should cast string → int once at the boundary. The less future maintainers have to think about what the data type might be, the better.

**Comment by noisysocks** [comment]
OK I'm happy with the code 👍 Let's do a final test 😀

**Comment by noisysocks** [comment]
Only small issue I'm noticing is that this help text doesn't change when you switch between Auto and Manual mode:

Screenshot 2024-06-25 at 12 58 30

Other than that I think we're okay to merge. The new features seem relatively functional when the experiment is turned on (we'll obviously iterate) and nothing seems to be different when the experiment is turned off.

**Comment by andrewserong** [comment]
Just took this for a spin, and it's looking very cool!

I found one other difference when the experiment is switched off. With this PR applied, the columns input field stretches across horizontally, rather than including a range control. Is that intentional?

Before

image

After

image

In terms of the experiment itself, I found I could sometimes unintentionally wind up with overlapping rows if I played around with moving grid items. Using the movers, I could get into a situation where the top row is fixed in the top row's position, and then I go to move the bottom row up, thinking that it'll nudge everything else to be beneath it. The result winds up being overlapping items instead of things moving out of the way. Not a blocker at all since it's part of the experiment, but just thought I'd share in case it's something to look into. It's a bit difficult to explain, so let me know if you'd like any extra info, and I'll see if I can explain it better.

The following video may not be very useful, it's mostly me getting a bit confused while using the movers 😅

https://github.com/WordPress/gutenberg/assets/14988353/e5515f83-de19-4678-88ce-c69a9afd30b8

Definitely not a blocker to landing, once the help text is fixed up!

File: packages/block-editor/src/components/grid-visualizer/use-grid-layout-sync.js
Comment (noisysocks) [inline packages/block-editor/src/components/grid-visualizer/use-grid-layout-sync.js:13]: Thinking out loud, it might make sense to trigger all of this logic from the control that toggles between Manual and Auto. That way we're not abusing `useEffect`.

Can change this later. Just something to think about

**Comment by tellthemachines** [comment]
Thanks for testing @andrewserong!

> With this PR applied, the columns input field stretches across horizontally, rather than including a range control. Is that intentional?

I don't think it's fully intentional as much as a side-effect of adding the Rows input when in Manual mode. I'm not sure if we should keep the range control when only Columns is present and remove it when we have Columns and Rows? In any case, we should probably revert the change at least when the experiment is disabled.

> Using the movers, I could get into a situation where the top row is fixed in the top row's position, and then I go to move the bottom row up, thinking that it'll nudge everything else to be beneath it. The result winds up being overlapping items instead of things moving out of the way.

This is deliberate 😅 following @jasmussen's request [above](https://github.com/WordPress/gutenberg/pull/61025#issuecomment-2180283919) to allow overlapping blocks. It can be a bit disconcerting if both blocks are exactly the same size, and maybe there's something we can do visually signal there's a hidden block, but the goal is to be able to put blocks on top of other blocks in manual mode!

File: packages/block-editor/src/components/grid-visualizer/use-grid-layout-sync.js
Comment (tellthemachines) [reply packages/block-editor/src/components/grid-visualizer/use-grid-layout-sync.js:13]: Yeah that would be worth looking into!

**Review by tellthemachines — APPROVED** [review]
Everything's working OK in testing now, let's get this in!

**Pull request merged**
